File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_sgd_vs_pytorch.cpp
/**
* @file test_sgd_vs_pytorch.cpp
* @brief Compares OwnTensor's SGD optimizer variations against PyTorch.
*
* Tests: Vanilla SGD, SGD+Momentum, SGD+WeightDecay, SGD+Momentum+WeightDecay
* Uses same inputs as Adam test for consistency.
*/
#include <iostream>
#include <iomanip>
#include <cmath>
#include <vector>


#include "TensorLib.h"


using namespace OwnTensor;


// Print tensor values to stdout with 7 decimal precision
void print_tensor(const std::string& name, const Tensor& t, int max_elements = 16) {
   Tensor tc = t.to_cpu();
   const float* d = tc.data<float>();
   int n = std::min(max_elements, (int)tc.numel());
  
   std::cout << name << " [" << tc.numel() << " elements, showing first " << n << "]:" << std::endl;
   for (int i = 0; i < n; ++i) {
       std::cout << std::fixed << std::setprecision(7) << d[i];
       if (i < n - 1) std::cout << ", ";
       if ((i + 1) % 8 == 0) std::cout << std::endl;
   }
   if (n % 8 != 0) std::cout << std::endl;
}


void test_sgd(const std::string& name, float lr, float momentum, float weight_decay) {
   std::cout << "\n========================================" << std::endl;
   std::cout << "=== SGD (" << name << "): OwnTensor ===" << std::endl;
   std::cout << "========================================\n" << std::endl;
  
   std::cout << "Hyperparameters:" << std::endl;
   std::cout << "  lr = " << lr << std::endl;
   std::cout << "  momentum = " << momentum << std::endl;
   std::cout << "  weight_decay = " << weight_decay << std::endl;
   std::cout << std::endl;
  
   // Create weight tensor: [0.1, 0.2, ..., 1.6]
   TensorOptions opts = TensorOptions()
       .with_dtype(Dtype::Float32)
       .with_req_grad(true);
  
   Tensor W = Tensor::zeros(Shape{{4, 4}}, opts);
   float* w_data = W.data<float>();
   for (int i = 0; i < 16; ++i) {
       w_data[i] = 0.1f * (i + 1);
   }
   W.set_requires_grad(true);
  
   std::cout << "--- Initial Weights ---" << std::endl;
   print_tensor("W_initial", W);
  
   // Fixed gradient [1.0, 1.0, ..., 1.0]
   Tensor grad = Tensor::ones(Shape{{4, 4}}, TensorOptions().with_dtype(Dtype::Float32));
  
   std::cout << "\n--- Gradient (fixed) ---" << std::endl;
   print_tensor("grad", grad);
  
   // Create SGD optimizer
   nn::SGDOptimizer sgd({W}, lr, momentum, weight_decay);
  
   // Set gradient and step
   W.set_grad(grad);
   sgd.step();
  
   std::cout << "\n--- After Step 1 ---" << std::endl;
   print_tensor("W", W);
}


int main() {
   std::cout << "================================================================" << std::endl;
   std::cout << "    OwnTensor SGD Optimizer vs PyTorch Accuracy Test" << std::endl;
   std::cout << "================================================================" << std::endl;
  
   float lr = 0.1f;
  
   // Test all SGD variations
   test_sgd("Vanilla", lr, 0.0f, 0.0f);
   test_sgd("Momentum", lr, 0.9f, 0.0f);
   test_sgd("WeightDecay", lr, 0.0f, 0.01f);
   test_sgd("Momentum+WeightDecay", lr, 0.9f, 0.01f);
  
   std::cout << "\n================================================================" << std::endl;
   std::cout << "    Now run: python3 test_sgd_vs_pytorch.py" << std::endl;
   std::cout << "================================================================" << std::endl;
  
   return 0;
}

File: DTensor/Tensor-Implementations/Tests/OptimizerTests/test_sgd_vs_pytorch.py

#!/usr/bin/env python3
"""
test_sgd_vs_pytorch.py


PyTorch reference for SGD optimizer variations.
Tests: Vanilla SGD, SGD+Momentum, SGD+WeightDecay, SGD+Momentum+WeightDecay
"""
import torch


def print_tensor(name, t, max_elements=16):
   """Print tensor values with 7 decimal precision."""
   data = t.detach().cpu().flatten().numpy()
   n = min(max_elements, len(data))
  
   print(f"{name} [{len(data)} elements, showing first {n}]:")
   for i in range(n):
       print(f"{data[i]:.7f}", end="")
       if i < n - 1:
           print(", ", end="")
       if (i + 1) % 8 == 0:
           print()
   if n % 8 != 0:
       print()


def test_sgd(name, lr, momentum, weight_decay):
   """Test SGD optimizer with specified parameters."""
   print("\n" + "=" * 50)
   print(f"=== SGD ({name}): PyTorch Reference ===")
   print("=" * 50 + "\n")
  
   print("Hyperparameters:")
   print(f"  lr = {lr}")
   print(f"  momentum = {momentum}")
   print(f"  weight_decay = {weight_decay}")
   print()
  
   # Create weight tensor: [0.1, 0.2, ..., 1.6]
   W = torch.zeros(4, 4, dtype=torch.float32, requires_grad=True)
   with torch.no_grad():
       for i in range(16):
           W.view(-1)[i] = 0.1 * (i + 1)
  
   print("--- Initial Weights ---")
   print_tensor("W_initial", W)
  
   # Fixed gradient [1.0, 1.0, ..., 1.0]
   grad = torch.ones(4, 4, dtype=torch.float32)
   print("\n--- Gradient (fixed) ---")
   print_tensor("grad", grad)
  
   # Create SGD optimizer
   optimizer = torch.optim.SGD([W], lr=lr, momentum=momentum, weight_decay=weight_decay)
  
   # Set gradient and step
   optimizer.zero_grad()
   W.grad = grad.clone()
   optimizer.step()
  
   print("\n--- After Step 1 ---")
   print_tensor("W", W)


def main():
   print("=" * 60)
   print("    PyTorch SGD Optimizer Reference")
   print("=" * 60)
  
   lr = 0.1
  
   # Test all SGD variations
   test_sgd("Vanilla", lr, momentum=0.0, weight_decay=0.0)
   test_sgd("Momentum", lr, momentum=0.9, weight_decay=0.0)
   test_sgd("WeightDecay", lr, momentum=0.0, weight_decay=0.01)
   test_sgd("Momentum+WeightDecay", lr, momentum=0.9, weight_decay=0.01)
  
   print("\n" + "=" * 60)
   print("    Done - Compare with OwnTensor output")
   print("=" * 60)


if __name__ == "__main__":
   main()
